### ISSUE TITLE
[Discussion] Add functionality to retry the webhook execution after being rate limited.

### ISSUE THREAD

**Issue by novasz** [lovvskillz/python-discord-webhook#47]
After inspection of the python-discord-webhook code base, I could not find any functionality that allowed for the webhook to be executed again if the user has received a status_code of 429 and is being rate limited by Discord. It seems like that any sort of retry based error handling is down to the user and there is currently no documentation on how to error handle being rate limited.

I have made my own implementation to handle being rate limited and retry to send the webhook again after a specific period of time (dependent on the webhook response sent after being rate limited), however, before I submit a pull request for this I would like to know if people would find this to be a useful feature to add to the package or would people rather just make their own implementation for handling this type of error.

**Comment by vremes** [comment]
Since this module is based on `requests`, i think it would also be nice to also add `timeout` arg/parameter to `DiscordWebhook` class.

When sending HTTP requests using `requests`, you can specify `timeout` parameter, example:

```python
# Raises requests.exceptions.Timeout exception after 5 seconds if server does not respond in 5 seconds.
response = requests.post(url, json=self.json, proxies=self.proxies, params={'wait':True}, timeout=5)
```

**Comment by r1bnc** [comment]
+1

**Comment by novasz** [comment]
If you send too many webhooks in a short period of time then Discord will temporarily disable you from sending more webhooks. They will send you a error code of 429 in the response and how much time you have to wait before you can send another webhook. I will probably send a pull request for the functionality to auto manage being rate limited, but for now, here is a quick way and inelegant you can manage being rate limited. 

```
    if webhook_execute.status_code in [200, 204]:
        print(Webhook successfully sent!')
        return

    while webhook_execute.status_code == 429:
        errors = json.loads(webhook_execute.content.decode('utf-8'))
        wh_sleep = (int(errors['retry_after']) / 1000) + 0.15
        time.sleep(wh_sleep)
        print(f'Webhook rate limited: sleeping for {wh_sleep} seconds...')
        webhook_execute = webhook.execute()
        if webhook_execute.status_code in [200, 204]:
            print(f'Webhook successfully sent!')
            return

```

Basically as long as you are getting rate limited (status 429) then it will wait for the time specified by the webhook response object and keep on attempting to resend the webhook.

**Comment by r1bnc** [comment]
I have been rate limited, I upload chat messages and the stream of messages kinda bugged out and made fast webhook post requests leading to being rate imited. Can I ask @novasz permission to use his code above, i gave him credit. Since, the chat messages are not so important, I just removed the `while` code and changed it so the program will wait until the rate limit is over and move onto newer messages to upload.

**Comment by novasz** [comment]
Yeah of course- feel free to use the above code. When I get some time I will submit a pull request for it anyway, so that everyone can benefit from it.

The reason behind needing a while loop is that if you are sending multiple webhooks (either through multiple threads or multiple programs) that are using the same webhook url, then it may be the case that once the rate limit is over the webhooks try to execute again, but Discord adds another rate limit and maybe only a few of those webhooks that were waiting for the rate limit go through, but the rest will not go through and will be lost. That is why it is important to constantly check if you are being rate limited before abandoning the execution of the webhook.

Adding a time.sleep(1) will help, but I would assume it would still be possible to be rate limited and in some cases I have seen rate limit wait times that are much longer than 1 second.

**Comment by r1bnc** [comment]
Thanks, I use the webhook to log data from in real time, so if it got rate limited I decided to abandon the other data and just upload the newer ones to match the timestamp. I also added the n second delay, i've set it to one and will monitor about rate limits the entire day.

**Comment by novasz** [comment]
It is better to make the sleep time variable when waiting for the rate limit cool-down period. You can get the time in seconds by doing the following:

```
errors = json.loads(webhook_execute.content.decode('utf-8'))
sleep_time = (int(errors['retry_after']) / 1000)
```

But if the time.sleep(1) works for your use case then by all means use it.

**Comment by r1bnc** [comment]
I use both of them :D

**Comment by r1bnc** [comment]
Hmm. The webhook stopped uploading data, no ratelimits no errors. I'm checking if it is my code,

**Comment by r1bnc** [comment]
Wait, is the `retry_after` already in seconds? 
Check: https://discord.com/developers/docs/topics/rate-limits

**Comment by lovvskillz** [comment]
This function is available in `0.14.0`.
